File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
private:
    stack<long long> st;
    long long mini;
public:
    MinStack() {
        mini = LLONG_MAX;
    }
    
    void push(int value) {
        if(st.empty()) {
            mini = value;
            st.push(value);
        }
        else if(value < mini) {
            st.push(2LL * value - mini);
            mini = value;
        }
        else {
            st.push(value);
        }
    }
    
    void pop() {
        if(st.empty()) return;
        if(st.top() < mini) {
            long long val = st.top();
            mini = 2 * mini - val;
        }

        st.pop();

        if(st.empty()) {
            mini = LLONG_MAX;
        }
    }
    
    int top() {
        if(st.empty())
            return -1;

        if(st.top() < mini) 
            return mini;

        return st.top();
    }
    
    int getMin() {
        return mini;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(value);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack/0155-min-stack.java
class MinStack {
    Stack<Long> st;
    long mini;

    public MinStack() {
        st = new Stack<>();
        mini = Long.MAX_VALUE;
    }
    
    public void push(int value) {
        if(st.isEmpty()) {
            st.push((long)value);
            mini = value;
        }
        else if(value < mini) {
            long newVal = (2L * value) - mini;
            st.push(newVal);
            mini = value;
        }
        else {
            st.push((long)value);
        }
    }
    
    public void pop() {
        if(st.isEmpty()) return;

        if(st.peek() < mini) {
            long newVal = st.peek();
            mini = (2 * mini) - newVal;
        }

        st.pop();

        if(st.isEmpty()) {
            mini = Long.MAX_VALUE;
        }
    }
    
    public int top() {
        if(st.empty()) return -1;

        if(st.peek() < mini) {
            return (int) mini;
        }

        return st.peek().intValue();
    }
    
    public int getMin() {
        return (int)mini;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(value);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */